### PR TITLE
Use of local data for calendar events instead of EventKit

### PR DESCRIPTION
**Pull request by Einh06**
## Changes overview
This PR changes the behavior of how course events are stored and processed, removing the dependency on EventKit to show events in the app, and use our own system.
We still have an opt-in mechanism to synchronize events with Apple iCal that can be found in the settings.

### Event Data
The CoreData entity `EventData` has been changed completely, from simply storing IDs (Remote, EventKit ID, and Course ID) to contain all event information received from the remove server.
This class is only used as storage, all user level usage to these information are done via `CalendarEvent`

### CalendarEvent
This is the user facing struct to process events. It has user friendly representation of `RecurrenceRule`, lots of convenience regarding next event calculation (through `EventSequence`). 
`EventDateIterator` is where the logic to calculate the next event's date is.

### CalendarEventHelper
This is a struct filled with convenience method to fetch and parse events, as well as pushing data to EventKit (located in CalendarEventHelper+EventKit).

## Detail on the event date calculation with recurrence rule
This is the meat of this PR.
The simple case is where an event doesn't have any recurrence rule associated with it. The event is kept as is, and won't repeat.

When it comes to recurrence rule there are 3 things to consider:
- when the event starts
- how the event repeats
- when the event ends

We gonna go through them from the simplest to the most complex

### When the event ends
If the `endDate` field is specified, the event ends at that date, otherwise it is infinitely recurring.

### How it repeats
There are 2 fields that directly affect how it repeats. The `frequency` and `interval`.
The frequency represent whether the event is a daily, weekly, monthly or yearly event. 
The interval is how many occurrence of that frequency until the next event. For example, a weekly event with an interval of 2 represent a biweekly event.

### How it starts
This is the trickiest part of it all. 
The startDate and endDate are not necessarily representing the start date. It does specify the start time and end time of the event, not necessarily the date. To find the date, it is necessary to use the dayOfWeek to figure it out. 
If the start date is Monday 28.08.2017, but the dayOfWeek specify Tuesday, the real start date is moved forward to match a Tuesday in that case (29.08.2017). 
As another example. If the start the Friday 01.09.2017, but the day of the week is Monday, we move the start date forward to match Monday (in that case 04.09.2017)

### CODE REVIEW

File: schulcloud/CalendarEvent.swift [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            case tuesday = "TU"
+            case wednesday = "WE"
+            case thursday = "TH"
+            case friday = "FR"
+            case saturday = "SA"
+            case sunday = "SU"
+        }
+    }
+    
+    init(id: String, title: String, description: String, location: String, startDate: Date, endDate: Date, rule: RecurrenceRule?, coreDataID: NSManagedObjectID?, ekEventID: String?) {
```
Comment (mathebox) [inline schulcloud/CalendarEvent.swift]: can this be private?

File: schulcloud/CalendarEvent.swift [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            
+            let internalEventWeekDay = startDate.weekday
+            
+            // we manually assign weekday indexes for each day, sunday being exception because in a german week sunday is the last day and not the first
+            // to make things easy we simply assign sunday 8 (1 + 7days)
+            
+            var dayOfWeekIndex : Int = {
+                switch rule!.dayOfTheWeek {
+                case .sunday:
+                    return 8 // is because 1
```
Comment (mathebox) [inline schulcloud/CalendarEvent.swift]: I don't get the comment.

File: schulcloud/CalendarEventHelper+EventKit.swift [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+    private struct Keys {
+        static let isSynchonized = "org.schul-cloud.calendar.eventKitIsSynchonized"
+        static let calendarIdentifier = "org.schul-cloud.calendar.identifier"
+    }
+    
+    struct EventKitSettings {
+        
+        static var current : EventKitSettings = EventKitSettings()
+        var isSynchonized : Bool {
```
Comment (mathebox) [inline schulcloud/CalendarEventHelper+EventKit.swift]: I would prefer `shouldSynchronize`. Please also change to name of the key.

File: schulcloud/EventData+.swift [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            return true
+        case "WEEKLY":
+            return true
+        case "MONTHLY":
+            return true
+        case "YEARLY":
+            return true
+        default:
+            return false
+        }
```
Comment (mathebox) [inline schulcloud/EventData+.swift]:
```swift
return ["DAILY", "WEEKLY", "MONTHLY", "YEARLY"].contains(remoteString)
```
could do this in one line

File: schulcloud/EventData+.swift [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            return true
+        case "FR":
+            return true
+        case "SA":
+            return true
+        case "SU":
+            return true
+        default:
+            return false
+        }
```
Comment (mathebox) [inline schulcloud/EventData+.swift]: See prev comment

File: schulcloud/SettingsViewController.swift [hunk 6]
```diff
@@ -32,5 +42,58 @@ class SettingsViewController: UITableViewController {
...
+                self.showErrorAlert(message: error.localizedDescription)
+                currentEventKitSettings.isSynchonized = true
+                switchView.isOn = true
+            }
+        }
+    }
+    
+    private func showErrorAlert(message: String) {
+        let alert = UIAlertController(title: "Somthing went wrong", message: message, preferredStyle: .alert)
+        self.present(alert, animated: true)
```
Comment (mathebox) [inline schulcloud/SettingsViewController.swift]: The alert can't be dismissed. There is no action.

File: schulcloud/SettingsViewController.swift [hunk 7]
```diff
@@ -32,5 +42,58 @@ class SettingsViewController: UITableViewController {
...
+                //TODO: Show error on why we could not delete the calendar
+                self.showErrorAlert(message: error.localizedDescription)
+                currentEventKitSettings.isSynchonized = true
+                switchView.isOn = true
+            }
+        }
+    }
+    
+    private func showErrorAlert(message: String) {
+        let alert = UIAlertController(title: "Somthing went wrong", message: message, preferredStyle: .alert)
```
Comment (mathebox) [inline schulcloud/SettingsViewController.swift]: Typo: Something

**Review by mathebox — CHANGES_REQUESTED** [review]
Look really good to me! Good job!

I have only some requests to make this an excellent PR (additionally to my inline comments):
- for some (new) methods the body starts with an empty line. Please remove these lines.
- It is possible to revoke the granted calendar permission via the Settings app. Make sure that the app check the calendar permission every time when EventKit is used. This includes updating the state of the switch in the Settings tab. The calendar can't be synchronize if the app hasn't the permission to do so. One path to trigger an error:
  1. Enable the calendar synchronization (grant access to calendar)
  2. disable calendar permission via settings app
  3. go back to schul-cloud app (switch is still enabled but shouldn't be)
  4. disable calendar sync (everything looks good)
  5. enable calendar sync (alert: somethings went wrong)

File: schulcloud/EventData+.swift (hunk 4 above)
Comment (Einh06) [reply schulcloud/EventData+.swift]: Oh yes! Much cleaner! Will do!

File: schulcloud/CalendarEvent.swift (hunk 1 above)
Comment (Einh06) [reply schulcloud/CalendarEvent.swift]: I need this in the tests.

File: schulcloud/CalendarEvent.swift (hunk 2 above)
Comment (Einh06) [reply schulcloud/CalendarEvent.swift]: I will update this.

File: schulcloud/CalendarEventHelper+EventKit.swift (hunk 3 above)
Comment (Einh06) [reply schulcloud/CalendarEventHelper+EventKit.swift]: That's more descriptive of what is going indeed. Will update.

File: schulcloud/SettingsViewController.swift (hunk 6 above)
Comment (Einh06) [reply schulcloud/SettingsViewController.swift]: Indeed, fixing that.

**Review by mathebox — APPROVED** [review]

**Pull request merged**